drivers: modem: introduce modem context helper and layers

mike-scott opened the pull request:
Initial support for modems in Zephyr use the following driver model:
- Main portions of code live in the modem specific driver.  This includes internal socket management, command parsing, etc.
- They leverage a UART-based modem receiver helper to gather data.
- Interface with Zephyr networking via net_context offload APIs.
    
This implementation was good enough to kick start interest in supporting modem usage in Zephyr, but lacks future scalability:
- The net_context offload APIs don't allow for operations such as offloaded DNS, SSL/TLS and other HW specific features.
- Since most of the code lives within the modem drivers, it's very hard for the Zephyr community to improve the driver layer over time.  Bugs found in 1 driver probably affect others due to copy/paste method of development.
- Lack of abstraction for different modem interfaces and command handlers makes it impossible to write a "dummy" layer which could be used for testing.
- Lack of centralized processing makes implementing low power modes and other advanced topics more difficult.
    
Introducing the modem context helper driver and sub-layers:
- modem context helper acts as an umbrella for several configurable layers and exposes this data to externals such as the modem shell.  Included in the helper is GPIO pin config functions which are currently duplicated in most drivers.
- modem interface layer: this layer sits on the HW APIs for the peripheral which communicates with the modem.  Users of the modem interface can handle data via read/write functions.  Individual modem drivers can select from (potentially) several modem interfaces.
- modem command parser layer: this layer communicates with the modem interface and processes the data for use by modem drivers.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/17922

mike-scott committed:
drivers: modem: context helper: introduce modem context helper driver

Initial support for modems in Zephyr use the following driver model:
- Main portions of code live in the modem specific driver.
  This includes internal socket management, command parsing, etc.
- They leverage a UART-based modem receiver helper to gather data.
- Interface with Zephyr networking via net_context offload APIs.

This implementation was good enough to kick start interest in
supporting modem usage in Zephyr, but lacks future scalability:
- The net_context offload APIs don't allow for operations such
  as offloaded DNS, SSL/TLS and other HW specific features.
- Since most of the code lives within the modem drivers, it's
  very hard for the Zephyr community to improve the driver layer
  over time.  Bugs found in 1 driver probably affect others due
  to copy/paste method of development.
- Lack of abstraction for different modem interfaces and command
  handlers makes it impossible to write a "dummy" layer which
  could be used for testing.
- Lack of centralized processing makes implementing low power modes
  and other advanced topics more difficult.

Introducing the modem context helper driver and sub-layers:
- modem context helper acts as an umbrella for several configurable
  layers and exposes this data to externals such as the modem shell.
  Included in the helper is GPIO pin config functions which are
  currently duplicated in most drivers.
- modem interface layer: this layer sits on the HW APIs for the
  peripheral which communicates with the modem.  Users of the modem
  interface can handle data via read/write functions.  Individual
  modem drivers can select from (potentially) several modem
  interfaces.
- modem command parser layer: this layer communicates with the
  modem interface and processes the data for use by modem drivers.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/17922

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: interface: introduce UART interface driver layer

The UART-based modem interface layer implements the modem context
interface for Zephyr's UART APIs.  This driver closely resembles
the existing modem receiver, but conforming to the modem interface
agreements.

Example modem driver setup code looks like this:

/* create modem context object */
static struct modem_context mctx;

/* create uart interface data object and buffers */
static struct modem_iface_uart_data iface_data;
static u8_t iface_isr_buf[MDM_RECV_BUF_SIZE];
static u8_t iface_rb_buf[MDM_MAX_DATA_LENGTH];

iface_data.isr_buf = &iface_isr_buf[0];
iface_data.isr_buf_len = sizeof(iface_isr_buf);
iface_data.rx_rb_buf = &iface_rb_buf[0];
iface_data.rx_rb_buf_len = sizeof(iface_rb_buf);
ret = modem_iface_uart_init(&mctx.iface, &iface_data,
			    UART_DEV_NAME);

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: cmd handler: introduce cmd handler driver layer

This is a generic command handler implementation which uses the
supplied modem interface to process incoming data and hand it
back to the modem driver via callbacks defined for:
- modem responses
- unsolicited messages
- specified handlers for current operation

The individual modem drivers define functions as command handlers
via the MODEM_CMD_DEFINE() macro.

To use these handlers, a modem operation defines a series of
modem_cmd structures and passes them to the modem_cmd_send()
function.  The modem_cmd includes data for:
- a matching string for when to execute the handler
- # of parameters to parse after the matching string
- delimeters for the parameters

Example modem driver setup code looks like this:

/* create modem context object */
static struct modem_context mctx;

/* net_buf receive pool */
NET_BUF_POOL_DEFINE(mdm_recv_pool, MDM_RECV_MAX_BUF,
		    MDM_RECV_BUF_SIZE, 0, NULL);

/* modem cmds */
static struct modem_cmd_handler_data cmd_handler_data;
static u8_t cmd_read_buf[MDM_RECV_BUF_SIZE];
static u8_t cmd_match_buf[MDM_RECV_BUF_SIZE];

/* modem response handlers */
static struct modem_cmd response_cmds[] = {
	MODEM_CMD("OK", on_cmd_ok, 0U, ""),
	MODEM_CMD("ERROR", on_cmd_error, 0U, ""),
	MODEM_CMD("+CME ERROR: ", on_cmd_exterror, 1U, ""),
};

/* unsolicited handlers */
static struct modem_cmd unsol_cmds[] = {
	MODEM_CMD("+UUSOCL: ", on_cmd_socknotifyclose, 1U, ""),
	MODEM_CMD("+UUSORD: ", on_cmd_socknotifydata, 2U, ","),
	MODEM_CMD("+UUSORF: ", on_cmd_socknotifydata, 2U, ","),
	MODEM_CMD("+CREG: ", on_cmd_socknotifycreg, 1U, ""),
};

/* setup cmd handler data */
cmd_handler_data.cmds[CMD_RESP] = response_cmds;
cmd_handler_data.cmds_len[CMD_RESP] = ARRAY_SIZE(response_cmds);
cmd_handler_data.cmds[CMD_UNSOL] = unsol_cmds;
cmd_handler_data.cmds_len[CMD_UNSOL] = ARRAY_SIZE(unsol_cmds);
cmd_handler_data.read_buf = &cmd_read_buf[0];
cmd_handler_data.read_buf_len = sizeof(cmd_read_buf);
cmd_handler_data.match_buf = &cmd_match_buf[0];
cmd_handler_data.match_buf_len = sizeof(cmd_match_buf);
cmd_handler_data.buf_pool = &mdm_recv_pool;
cmd_handler_data.alloc_timeout = BUF_ALLOC_TIMEOUT;
ret = modem_cmd_handler_init(&mctx.cmd_handler, &cmd_handler_data);

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: introduce socket helper layer

Many modems implement socket-based APIs to manage data connections.
This layer provides much of the groundwork for keeping track of
these "sockets" throughout their lifecycle (from the initial offload
API calls through the command handler call back layers):
- structure for holding socket data like IP protocol, destination,
  source and incoming packet sizes
- configuration to note modem starting socket id and number of
  sockets
- methods to get/put socket structs from/to the pool
- function to update the # and size of packets in the modem receive
  queue
- prebuilt modem_socket_poll() method for socket offload poll() API

Example modem driver setup code looks like this:

/* socket data */
static struct modem_socket_config socket_config;
static struct modem_socket sockets[MDM_MAX_SOCKETS];

static int modem_init(struct device *dev)
{
  ...
  /* setup socket config */
  socket_config.sockets = &sockets[0];
  socket_config.sockets_len = ARRAY_SIZE(sockets);
  socket_config.base_socket_num = 0;
  ret = modem_socket_init(&socket_config);
  ...
}

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: ublox SARA convert to modem context

Let's convert the SARA modem to use the more generic modem context
layers so that we don't maintain a lot of what should be shared code.

This conversion includes:
- modem context as the helper umbrella
- uart modem interface layer
- generic command handler layer
- modem socket helper
- move from net_context offload API to socket offload API

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: ublox-sara-r4: Support SARA-U2 modems, sense VINT

This adds support for SARA-U2 modems. They have different timings on
the PWR_ON pin, don't support AT+CESQ and require a manual GPRS
connection setup.

The VINT pin is used as a more reliable and faster way to power on the
modem.

Based on work by Göran Weinholt <goran.weinholt@endian.se>

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Including the following who have shown interest in modems:
@weinholtendian
@WilliamGFish
@rerickson1
@cybertale
@Oanerer
@Swij

ulfalizer commented on drivers/modem/Kconfig:
```diff
@@ -30,15 +30,66 @@ config MODEM_RECEIVER
 	  Maximum number of modem receiver contexts to handle.  For most
 	  purposes this should stay at 1.
 
+endif # MODEM_RECEIVER
```
Is this change in case other symbols that depend on `MODEM_RECEIVER` get added later?

The `depends on` is simpler otherwise.

mike-scott replied:
Yep.  Will change back.

mike-scott commented:
My original goal was to get this in for Zephyr 2.0.  But, I had a lot of work distractions so this came in very late.  As this is a new driver helper, it doesn't affect anyone's existing driver so maybe there is still time.

By getting it into v2.0 we can establish a baseline for new modem drivers to use, rather than end up with a lot of copy/paste code.

There is a lot of room for improvement over time, but that should come later:
- power management
- dummy modem driver + modem interface for testing
- etc

mike-scott commented:
Also: my ELC NA 2019 presentation will center around these changes

mike-scott commented:
Ugh.  Fixing checkpatch issues.  :)

ulfalizer commented on drivers/modem/Kconfig:
```diff
@@ -30,15 +30,66 @@ config MODEM_RECEIVER
+	  purposes this should stay at 1.
+
+config MODEM_CONTEXT_VERBOSE_DEBUG
+	bool "Enable verbose debug in modem context"
```
Could skip the "Enable ..." parts here and below maybe, since it's clear in the `menuconfig` from the checkbox.

Something like

    bool "Verbose debug output in modem context"
    bool "UART-based modem interface"
    bool "Generic modem command handler"
    bool "u-blox SARA modem driver"
    ...

etc.

ulfalizer commented on drivers/modem/Kconfig:
```diff
@@ -30,15 +30,66 @@ config MODEM_RECEIVER
+	depends on SERIAL_SUPPORT_INTERRUPT
+	select UART_INTERRUPT_DRIVEN
+	select RING_BUFFER
+	bool "Enable UART-based modem interface"
```
Could move this line up to just below the `config MODEM_IFACE_UART` line.

ulfalizer commented on drivers/modem/Kconfig:
```diff
@@ -30,15 +30,66 @@ config MODEM_RECEIVER
+	bool "Enable UART-based modem interface"
+
+config MODEM_CMD_HANDLER
+	bool "Enable generic modem command handler"
```
Some help texts for these might be nice, to quickly point people to what to do next after they enable them.

ulfalizer reviewed (commented):
Just some subjective nits. Nothing obligatory.

mike-scott commented:
@ulfalizer Thank you for the review!

Posted v2 with diff here: https://pastebin.com/raw/gbrrK9x9

- Addressed comments
- Fixed checkpatch and Kconfig issues

mike-scott commented:
@ulfalizer Looks like I need to think about this "choice" between R4 and U2 variants of the modem.  It's breaking Kconfig.

mike-scott commented:
@ulfalizer specifically:
```
Error: Aborting due to non-whitelisted Kconfig warning 'warning: default on the choice symbol
MODEM_UBLOX_SARA_R4 (defined at boards/arm/particle_boron/Kconfig.defconfig:64,
drivers/modem/Kconfig.ublox-sara-r4:27) will have no effect, as defaults do not affect choice
symbols'. Note: If this warning doesn't point to an actual problem, you can add it to the whitelist
at the top of /home/scottml/fio/zephyr-master/zephyr/scripts/kconfig/kconfig.py.
```

This is caused by how I'm selecting MODEM_UBLOX_SARA_R4 in the particle_boron Kconfig.defconfig:
```
if MODEM

config UART_1_NRF_UARTE
	default y

config UART_1_NRF_FLOW_CONTROL
	default y

config MODEM_UBLOX_SARA
	default y

config MODEM_UBLOX_SARA_R4
	default y

config UART_INTERRUPT_DRIVEN
	default y

endif # MODEM
```

ulfalizer commented:
@mike-scott 
Might need to give the choice a name and define it in multiple locations.

There's some instructions in the [board porting guide](https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#configuring-choices) (kinda weird location, yeah).

mike-scott commented:
@ulfalizer thank you for the recommendation!  Pushed new version with Kconfig fixes.
Diff from v2: https://pastebin.com/raw/JrNTURX4

jukkar commented on drivers/modem/Kconfig:
```diff
@@ -39,6 +39,33 @@ config MODEM_RECEIVER_MAX_CONTEXTS
 	  Maximum number of modem receiver contexts to handle.  For most
 	  purposes this should stay at 1.
 
+config MODEM_CONTEXT
```
I suggest marking this as experimental, so that it is easier to modify the APIs etc if we find out that there are changes needed, so just add "[EXPERIMENTAL]" at the end of bool line below.

mike-scott replied:
Good idea!  Thanks.

jukkar commented on drivers/modem/modem_context.c:
```diff
@@ -0,0 +1,185 @@
+void modem_context_hexdump(const char *desc, const u8_t *packet, size_t length)
+{
+#if defined(CONFIG_MODEM_CONTEXT_VERBOSE_DEBUG)
+	char output[sizeof("xxxxyyyy xxxxyyyy")];
```
Why cannot we just use `LOG_HEXDUMP_DBG()` here?

mike-scott replied:
Mainly because I forgot it existed!  Fixing.

jukkar commented on drivers/modem/modem_context.c:
```diff
@@ -0,0 +1,185 @@
+
+#include "modem_context.h"
+
+static struct modem_context *contexts[CONFIG_MODEM_CONTEXT_MAX_NUM];
```
You have array of pointers here, where is the actual context allocated, did not spot it immediately.

mike-scott replied:
Users allocate their own `modem_context` like this in the ublox sara driver:
https://github.com/zephyrproject-rtos/zephyr/pull/18103/files#diff-cb3c7d77935fa5df6ccb74a1c5e050baR149

Then call `modem_context_register()` to store the modem_context in the master list exposing it to "externals" such as modem_shell:
https://github.com/zephyrproject-rtos/zephyr/pull/18103/files#diff-cb3c7d77935fa5df6ccb74a1c5e050baR1274

This is the `modem_context_register()` function:
https://github.com/zephyrproject-rtos/zephyr/pull/18103/files#diff-395ec129e305de98c66d00aef6c64373R165

Which calls an internal modem_context_get() function to assigns the modem_context pointer to the internal list of pointers:
https://github.com/zephyrproject-rtos/zephyr/pull/18103/files#diff-395ec129e305de98c66d00aef6c64373R142

jukkar replied:
Ok, thanks for clarification.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -176,1114 +128,516 @@ struct modem_iface_ctx {
+	struct modem_socket *sock;
 
-		hexdump(uart_buffer, bytes_read);
+	socket_id = atoi(argv[0]);
```
I just wonder if we should use strtol() instead as atoi() cannot figure out errors in the string.
So all the invocations of atoi() could be replaced by
```
	socket_id = strtol(argv[0], &endptr, 10);
	if (*endptr == '\0') {
		/* string ok */
	} else {
		/* not a number */
	}
```

mike-scott replied:
Will do.  Thanks for the recommendation.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1304,85 +660,30 @@ static void modem_reset(void)
-	ret = send_at_cmd(NULL, "AT+CGDCONT=1,\"IP\",\""
-				CONFIG_MODEM_UBLOX_SARA_R4_APN "\"",
-			  MDM_CMD_TIMEOUT);
+	ret = modem_cmd_handler_setup_cmds(&mctx.iface, &mctx.cmd_handler,
```
The function name contains a bit redundant data, would `modem_cmd_handler_setup()` be descriptive enough?

mike-scott replied:
Naming is hard :)  I'm not sure `modem_cmd_handler_setup()` is descriptive enough as that sounds like you're setting up a cmd_handler.

This function runs a bunch of setup commands via the command handler..  Not sure how to make that name better.

jukkar replied:
No worries, there was just many cmd chars in the function name. We can leave it like this, it is not really an issue.

mike-scott commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -0,0 +1,485 @@
+	u16_t offset, len;
+	bool quit_loop;
+
+	static u8_t *argv[CONFIG_MODEM_CMD_HANDLER_MAX_PARAM_COUNT];
```
Woops!  Nice catch.  That shouldn't be static.  It comes from the stack for this function where it's handed to a handler and then freed up again.

mike-scott replied:
(Was a last minute change moving it around to save memory)

mike-scott commented:
@jukkar updated with your recommendations.  Here's a diff from the previous version:
https://pastebin.com/raw/N93ZRD4G

- Added [EXPERIMENTAL] to MODEM_CONTEXT config
- Removed the "static" from argv/argc vars and cleaned up their handling a bit
- dropped `modem_context_hexdump()` in favor of `LOG_HEXDUMP_DBG()`.  Also had to re-work where the logging was done now that there is literally no pause in execution for logging (I was getting almost all single characters logged).
- Fixed `poll()` return codes for ETIMEDOUT and EAGAIN (don't return an error).
- Swapped atoi usage to strtol.  I set this up as a function + macro for readability due to line lengths.  If you don't like how I did that I can change as needed.
-

tbursztyka commented on drivers/modem/modem_iface_uart.c:
```diff
@@ -0,0 +1,150 @@
+	       uart_irq_rx_ready(ctx->iface.dev)) {
+		rx = uart_fifo_read(ctx->iface.dev,
+				    data->isr_buf, data->isr_buf_len);
+		if (rx > 0) {
```
I'd rather break if (rx == 0), less imbrication of if/if/if

mike-scott replied:
will fix.

mike-scott replied:
@tbursztyka actually, a break on (rx == 0) would change the current logic in 2 ways:
- `uart_fifo_read` can return an error (<0) so that needs to be handled as well
- current logic doesn't break the loop if rx is <=0, it re-checks `uart_irq_update` and `uart_irq_rx_ready`.

Instead I'll change to the following which preserves the logic:
```
               if (rx <= 0) {
                       continue;
               }
```

tbursztyka commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -0,0 +1,488 @@
+	}
+
+	/* process all of the data in the net_buf */
+	while (data->rx_buf) {
```
the while loop below is hard to digest. way too long, would benefit to get parts split into inline functions.

mike-scott replied:
I can break this up a bit.

jukkar commented:
>     * dropped `modem_context_hexdump()` in favor of `LOG_HEXDUMP_DBG()`.

You could keep the call to `modem_context_hexdump("RECV", data->read_buf, bytes_read);` in order to avoid extra `#if defined(CONFIG_MODEM_CONTEXT_VERBOSE_DEBUG)` in the code.
So something like this:
```
static void modem_context_hexdump(x, y z)
{
#if defined(CONFIG_MODEM_CONTEXT_VERBOSE_DEBUG)
		LOG_HEXDUMP_DBG(params);
#endif
}
```

tbursztyka commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -0,0 +1,488 @@
+/** @file
```
Choosing net_buf for such command handler was maybe a bit overkill. You always work with maximum size of buffer anyway.

mike-scott replied:
The nicest part of net_buf was the append / pull flexibility.   I can optimize this at a future date with a different max buffer primitive.  Thanks.

jukkar reviewed (approved):
LGTM

tbursztyka reviewed (approved):
Just 2-3 tiny comments. 
Choosing net_buf for the cmd handling was maybe overkill, but fine let's keep it.

Approving as the overall seems fine to me.

rlubos commented on drivers/modem/modem_socket.c:
```diff
@@ -0,0 +1,281 @@
+		}
+	}
+
+	ret = k_sem_take(&cfg->sem_poll, msecs);
```
The design here makes the `poll` function non-reentrant. If two different threads `poll` on two different sockets we'll end up with unexpected behavior - the higher priority thread will be unblocked, regardless on which socket it polled. I think we could live with such limitation though in the initial implementation, but this should be improved in the future.

And you could handle POLLOUT as well. We can consider the socket to be always writeable, and set `POLLOUT` in `revents` whenever it is requested, but this should be done. Otherwise, unaware `poll` users might block forever, waiting for a POLLOUT to be set.

mike-scott replied:
I will add FIXME comment here regarding reentrancy.   And set POLLOUT in revents with a comment as well.  Thanks.

rlubos commented on drivers/modem/modem_socket.c:
```diff
@@ -0,0 +1,281 @@
+
+#include <kernel.h>
+#include <net/socket_offload.h>
+#include <net/socket_offload_ops.h>
```
It seems that socket offloading is handled elsewhere in this design. You could remove the headers to avoid confusion.

mike-scott replied:
nice catch.  I attempted to move the offloaded ops into modem_socket at one point and left the headers.

rlubos reviewed (approved):
Looks like a fine groundwork for UART modem/socket integration.

ioannisg reviewed (commented):
Let's try to get this in 2.0 (and improve it until the release, if needed), if that's the consensus here

mike-scott commented:
@ioannisg its morning my time.  I'll have an update ready in 2ish hours + time for CI -- should be ready for today depending on timezones.

WilliamGFish commented:
Mike,
This is looking great, I still haven't found the offending code that causes the U2 version to encounter a FATAL error.

```
FATAL: ***** USAGE FAULT *****
FATAL:   Illegal use of the EPSR
FATAL: r0/a1:  0x00000000  r1/a2:  0x2000b1e0  r2/a3:  0x00000001
FATAL: r3/a4:  0x0000be77 r12/ip:  0x00000031 r14/lr:  0x00006bd1
FATAL:  xpsr:  0x60000000
FATAL: Faulting instruction address (r15/pc): 0x00008404
FATAL: >>> ZEPHYR FATAL ERROR 0: CPU exception
FATAL: Current thread: 0x20001320 (unknown)
FATAL: Halting system
```

The instruction address points to:
```
PRINT_BOOT_BANNER();
 00008404   LDR          R0, =0x0000FEA2                 ; [0x00008424]
 00008406   BL           printk                          ; 0x00009E6A
```

The change occurred between 1.14 and  July. I'll keep looking but am at a lose right now.

Billy..

WilliamGFish commented on drivers/modem/Kconfig:
```diff
@@ -39,6 +39,86 @@ config MODEM_RECEIVER_MAX_CONTEXTS
+
 config MODEM_SHELL
 	bool "Enable modem shell utilities"
 	select SHELL
```
Could we add a default to Y if MODEM selected, ie;

`default MODEM`

jukkar replied:
IMHO the shell should not be selected automatically, but user should manually select it if it is really needed. Shell uses memory and really depends on hw how useful it is.

mike-scott replied:
I agree: production quality software would not leave modem shell selected.

mike-scott commented:
Addressed comments from @tbursztyka @rlubos.  Here's the diff: https://pastebin.com/raw/JcVKSZaB

mike-scott commented:
@tbursztyka I chose not to mark the new functions in modem_cmd_handler.c as "inline".  I think "static" is enough and the compiler will normally perform that optimization if needed.  If you don't agree with that choice: I can update the patch.

mike-scott commented:
Summary of changes:

- introduced `process_cmd` and `find_cmd_match` functions into `modem_cmd_handler.c` to break up the loop in `cmd_handler_process`
- `out_len` in `cmd_handler_process` renamed to `match_len` to clarify and use it instead of the original `len` (to next CR/LF) in several cases.
- added a few more comments to the code for clarity
- changed `modem_iface_uart_isr` per @tbursztyka recommendation while keeping original logic intact
- dropped `net/socket_offload_ops.h` include from modem_socket.c
- added FIXME comment regarding `modem_socket_poll` limitations
- handle `POLLOUT` event in `modem_socket_poll` (exit early as we assume socket is always writable)

WilliamGFish commented:
RESOLVED -- FATAL error.
> 
> ```
> FATAL: ***** USAGE FAULT *****
> FATAL:   Illegal use of the EPSR
> FATAL: r0/a1:  0x00000000  r1/a2:  0x2000b1e0  r2/a3:  0x00000001
> FATAL: r3/a4:  0x0000be77 r12/ip:  0x00000031 r14/lr:  0x00006bd1
> FATAL:  xpsr:  0x60000000
> FATAL: Faulting instruction address (r15/pc): 0x00008404
> FATAL: >>> ZEPHYR FATAL ERROR 0: CPU exception
> FATAL: Current thread: 0x20001320 (unknown)
> FATAL: Halting system
> ```
> 
> The instruction address points to:
> 
> ```
> PRINT_BOOT_BANNER();
>  00008404   LDR          R0, =0x0000FEA2                 ; [0x00008424]
>  00008406   BL           printk                          ; 0x00009E6A
> z_sys_device_do_config_level(_SYS_INIT_LEVEL_APPLICATION);
>  0000840A   MOVS         R0, #3
>  0000840C   BL           z_sys_device_do_config_level    ; 0x00008248
> z_init_static_threads();
>  00008410   BL           z_init_static_threads           ; 0x000091F0
> main();
>  00008414   BL           main                            ; 0x00000470
> }
>  00008418   LDR          R2, =_main_thread_s             ; [0x00008428] =0x20001320
>  0000841A   LDRB         R3, [R2, #0xC]
>  0000841C   BIC          R3, R3, #1
>  00008420   STRB         R3, [R2, #0xC]
> init.c , Line 304
>  00008422   POP          {R3, PC}```
> 
> The change occurred between 1.14 and  July. I'll keep looking but am at a lose right now.
> 
> Billy..
> ```

Combination of configuration issue and a suspect power supply.
This combination causes hard fault;
CONFIG_HW_STACK_PROTECTION=y
CONFIG_MODEM_SHELL=y

Even not having CONFIG_MODEM_SHELL set will cause failures.

mike-scott commented:
@WilliamGFish Interesting!  What sample are you testing?

WilliamGFish commented:
@mike-scott  Its a BLE Mesh with MQTT to Blue-mix via Particle
The hardware protection settings was due to multiple issues when stressing the Mesh.

mike-scott commented:
@ioannisg I know we discussed merging this on Slack, but capturing here for historical reasons.  I feel we can probably merge this as today is the cut-off for feature freeze and the general approach of this patchset was approved..  However, I'd still like to get @tbursztyka and @rlubos to look at my final changes regarding their comments.  If there are other changes needed, I will do a follow-up bugfix patch.

mtahirbutt commented:
Hi, I just want reinitialize PPP and Offloading both when I hard switch off (power pin) the modem from the main application. Right now I just reinitialize gsm_init and modem_init when I reset from the main application. Any idea towards solution to this problem.
